### PR TITLE
Improve Excel parity for undefined, null & error inputs

### PR DESCRIPTION
**Pull request by pbadenski**
I matched behaviour of selected functions with Excel when undefined or error is passed as an argument. I haven't done this for all functions, but I tried to select these that I thought would be the most commonly used.

Related: #47, #48

### CODE REVIEW

**Comment by pbadenski** [comment]
@nicolashefti thoughts?

**Comment by nicolashefti** [comment]
@pbadenski thanks for the reminder. Sorry for the delay, I had a busy week. It's on my ToDo for the weekend.

File: lib/math-trig.js [hunk 1]
```diff
@@ -116,6 +131,12 @@ exports.AGGREGATE = function(function_num, options, ref1, ref2) {
 };
 
 exports.ARABIC = function(text) {
+  if (text === undefined || text === null) {
```
Comment (nicolashefti) [inline lib/math-trig.js:134]: utils.parseNumber?

File: lib/math-trig.js [hunk 2]
```diff
@@ -424,7 +580,11 @@ exports.FLOOR.PRECISE = exports.FLOOR.MATH;
 
 // adapted http://rosettacode.org/wiki/Greatest_common_divisor#JavaScript
 exports.GCD = function() {
-  var range = utils.parseNumberArray(utils.flatten(arguments));
+  var flatArguments = utils.flatten(arguments);
+  if (flatArguments.every(function (arg) { return arg === undefined || arg === null; })) {
```
Comment (nicolashefti) [inline lib/math-trig.js:584]: Can be checked in the existing loop

File: lib/math-trig.js [hunk 3]
```diff
@@ -461,8 +624,13 @@ exports.ISO = {
 };
 
 exports.LCM = function() {
+  var flatArguments = utils.flatten(arguments);
+  if (flatArguments.every(function (arg) { return arg === undefined || arg === null; })) {
```
Comment (nicolashefti) [inline lib/math-trig.js:628]: Can be checked in the existing loop?

File: lib/math-trig.js [hunk 4]
```diff
@@ -556,7 +760,11 @@ exports.MROUND = function(number, multiple) {
 };
 
 exports.MULTINOMIAL = function() {
-  var args = utils.parseNumberArray(utils.flatten(arguments));
+  var flatArguments = utils.flatten(arguments);
+  if (flatArguments.every(function (arg) { return arg === undefined || arg === null; })) {
```
Comment (nicolashefti) [inline lib/math-trig.js:764]: Can be checked in the existing loop

**Review by nicolashefti — COMMENTED** [review]
The refactoring is quite important. It would easier to digest it if you propose one PR per library part (logicial, trig, stats). The discussion would be more focused.

Some general comments:

- most of the time you go through all the parameters to check for an error. This can could be included in the main loop of the function to speed up things.

- There is a lot a duplicated code. Can you introduce something in the utils to make the parameters check less verbose?

- Sometime the code looks like this in the end:

```
  if (number instanceof Error) {
    return number;
  }

  number = utils.parseNumber(number);

  if (number instanceof Error) {
    return number;
  }
```

Can it be simplified?

- Same remark on this kind of code:

```
  if (number_x instanceof Error) {
    return number_x;
  }

  if (number_y instanceof Error) {
    return number_y;
  }

  number_x = utils.parseNumber(number_x);

  number_y = utils.parseNumber(number_y);

  if (utils.anyIsError(number_x, number_y)) {
    return error.value;
  }
```

**Comment by nicolashefti** [comment]
@pbadenski can you open a new PR with just for utils/common.js and lib/logical.js to start with?

**Comment by pbadenski** [comment]
Ok I will. I will copy your comments to that pull request and answer them there. I think probably better to move the conversation there sooner rather than later.

**Comment by pbadenski** [comment]
FYI I opened a pull request for lib/logical.js - once we resolve it, then I'll move to another. I'll make sure your feedback is copied over as we discuss the relevant code.

**Pull request closed without merging**